Include invoice requests in async payment onions

valentinewallace opened the pull request:
Title. Implements the sending-side of the final commit of https://github.com/lightning/bolts/pull/1149. 

Based on #3140.

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1186,7 +1192,7 @@ impl OutboundPayments {
 
 		onion_utils::set_max_path_length(
-			route_params, recipient_onion, keysend_preimage, best_block_height
+			route_params, recipient_onion, keysend_preimage, None, best_block_height
```
Need to pass in the invreq here.

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -951,20 +951,26 @@ impl OutboundPayments {
+		match outbounds.entry(payment_id) {
+			hash_map::Entry::Occupied(entry) => match entry.remove() {
+				PendingOutboundPayment::InvoiceReceived { .. } => {
+					outbounds.insert(payment_id, retryable_payment);
```
Hmm, I did this to avoid cloning the invoice request. It seemed necessary to fully move the payment out of the map to do so. Or do you mean avoid `remove`ing in the `::InvoiceReceived` case, and keep it in the `::StaticInvoiceReceived` case?

valentinewallace replied:
Check out the latest version!

valentinewallace commented on lightning/src/ln/msgs.rs:
```diff
@@ -2733,13 +2735,17 @@ impl<'a> Writeable for OutboundOnionPayload<'a> {
 				// We need to update [`ln::outbound_payment::RecipientOnionFields::with_custom_tlvs`]
 				// to reject any reserved types in the experimental range if new ones are ever
 				// standardized.
+				let invoice_request_tlv = invoice_request.map(|invreq| (77_777, invreq.encode())); // TODO: update TLV type once the async payments spec is merged
```
Oops 🤦‍♀️

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1048,6 +1049,11 @@ impl OutboundPayments {
 						keysend_preimage,
 						retry_strategy: *retry_strategy,
 						route_params,
+						invoice_request:
+							retryable_invoice_request
+							.take()
+							.ok_or(Bolt12PaymentError::UnexpectedInvoice)?
+							.invoice_request,
```
This may be `None` if the `InvoiceRequest` was already retried, so we'll need to have some other means of tracking whether or not to retry. I think this was mentioned in an earlier PR, so perhaps it will be handled in a follow-up?

valentinewallace replied:
Ah yeah, probably will do that in a follow-up. Looks like it's tracked here https://github.com/lightningdevkit/rust-lightning/issues/2298. Thanks for the reminder!

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2823,6 +2834,11 @@ mod tests {
 	}
 
+	fn invoice_request() -> InvoiceRequest {
+		let invreq_bytes = <Vec<u8>>::from_hex("00200101010101010101010101010101010101010101010101010101010101010101080203e80a00162102bb58b5feca505c74edc000d8282fc556e51a1024fc8e7d7e56c6f887c5c8d5f25821035be5e9478209674a96e60f1f037f6176540fd001fa1d64694770c56a7709c42cf040a31d68198578a5aa1fe57de5f6d33f89c1556752cb333dbb56ac727f751893804d3a26b42909f1952fb1db8080238f9476b829160387692df35ff85dfe72e2ed").un ...
+		Readable::read(&mut Cursor::new(&invreq_bytes[..])).unwrap()
```
Is there any way to create this programmatically?

valentinewallace commented:
Rebased due to conflicts, then pushed an additional typo fix:
```diff
diff --git a/lightning/src/events/mod.rs b/lightning/src/events/mod.rs
index bc7bbb641..823c5e2c2 100644
--- a/lightning/src/events/mod.rs
+++ b/lightning/src/events/mod.rs
@@ -566,7 +566,7 @@ pub enum PaymentFailureReason {
        /// routes - we tried the payment over a few routes but were not able to find any further
        /// candidate routes beyond those.
        ///
-       /// Also used for [`BlindedPathCreationFailed`] down downgrading to versions prior to 0.0.124.
+       /// Also used for [`BlindedPathCreationFailed`] when downgrading to versions prior to 0.0.124.
        RouteNotFound,
        /// This error should generally never happen. This likely means that there is a problem with
        /// your router.

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2834,9 +2832,28 @@ mod tests {
-		Readable::read(&mut Cursor::new(&invreq_bytes[..])).unwrap()
+	fn dummy_invoice_request() -> InvoiceRequest {
+		let secp_ctx = Secp256k1::new();
+		let recipient_pubkey = {
+			let secret_key = SecretKey::from_slice(&<Vec<u8>>::from_hex("4141414141414141414141414141414141414141414141414141414141414141").unwrap()).unwrap();
+			Keypair::from_secret_key(&secp_ctx, &secret_key).public_key()
+		};
+		let payer_keys = {
+			let secret_key = SecretKey::from_slice(&<Vec<u8>>::from_hex("4242424242424242424242424242424242424242424242424242424242424242").unwrap()).unwrap();
+			Keypair::from_secret_key(&secp_ctx, &secret_key)
+		};
```
Let's inline `recipient_pubkey()` and `payer_pubkey()` below instead.

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -2834,9 +2832,28 @@ mod tests {
+			.amount_msats(1000)
+			.build()
+			.unwrap()
+			.request_invoice(vec![0; 8], payer_keys.public_key()).unwrap()
+			.build()
+			.unwrap()
+			.sign(|message: &UnsignedInvoiceRequest|
+				Ok(secp_ctx.sign_schnorr_no_aux_rand(message.as_ref().as_digest(), &payer_keys))
+			)
```
This is interface being removed in #3264. Could use the derived version and `build_and_sign` instead?

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -948,20 +949,36 @@ impl OutboundPayments {
-			payment_params, entropy_source, best_block_height
-		);
-		match self.pending_outbound_payments.lock().unwrap().entry(payment_id) {
+		macro_rules! create_pending_payment {
```
Doesn't seem like we need a macros for this?

jkczyz commented on lightning/src/events/mod.rs:
```diff
@@ -560,11 +560,13 @@ pub enum PaymentFailureReason {
 	/// routes - we tried the payment over a few routes but were not able to find any further
 	/// candidate routes beyond those.
+	///
+	/// Also used for [`BlindedPathCreationFailed`] when downgrading to versions prior to 0.0.124.
```
CI sad. Need: 

```rust
/// [`BlindedPathCreationFailed`]: Self::BlindedPathCreationFailed
```

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -613,8 +613,9 @@ impl RouteParameters {
 		&mut self, recipient_onion: &RecipientOnionFields, is_keysend: bool, best_block_height: u32
 	) -> Result<(), ()> {
 		let keysend_preimage_opt = is_keysend.then(|| PaymentPreimage([42; 32]));
+		// TODO: no way to account for the invoice request here yet
```
Can't we pass it to `RouteParameters::set_max_path_length` as an `Option` parameter? Doesn't look like it is ever called internally, though. Is this mainly for use outside of `ChannelManager`?

valentinewallace replied:
We can though we'll need some way to clarify in the API that it should only be passed in for payments to a static invoice, which isn't part of LDK's public API yet. Not sure if it's worth addressing now. 

> Doesn't look like it is ever called internally, though. Is this mainly for use outside of ChannelManager?

Yep, just seems like a useful util on its own IIRC.

jkczyz commented on lightning/src/ln/msgs.rs:
```diff
@@ -2760,13 +2762,17 @@ impl<'a> Writeable for OutboundOnionPayload<'a> {
 				// We need to update [`ln::outbound_payment::RecipientOnionFields::with_custom_tlvs`]
 				// to reject any reserved types in the experimental range if new ones are ever
 				// standardized.
+				let invoice_request_tlv = invoice_request.map(|invreq| (77_777, invreq.encode())); // TODO: update TLV type once the async payments spec is merged
```
Will this type eventually be less than 2^16? If so, should we just pick something there and move this to `_encode_varint_length_prefixed_tlv` instead of using `custom_tlvs`?

valentinewallace replied:
My point of reference is mainly that eclair's approach has been to use experimental TLVs until the corresponding BOLT is close(r) to merge, e.g. for their experimental trampoline features. It does seem a little bit safer and might allow us to end up with a lower type since there wouldn't be risk of overlap with another feature?

jkczyz reviewed (commented):
LGTM other than some minor comments.

jkczyz reviewed (commented):
LGTM. Good to squash whenever.

valentinewallace commented:
I went ahead and squashed, the diff prior was: 

```diff
﻿﻿﻿diff --git a/lightning/src/ln/outbound_payment.rs b/lightning/src/ln/outbound_payment.rs
index c68209e8d..c3fde629a 100644
--- a/lightning/src/ln/outbound_payment.rs
+++ b/lightning/src/ln/outbound_payment.rs
@@ -954,7 +954,7 @@ impl OutboundPayments {
 		let onion_session_privs = match outbounds.entry(payment_id) {
 			hash_map::Entry::Occupied(entry) => match entry.get() {
 				PendingOutboundPayment::InvoiceReceived { .. } => {
-					let (retryable_payment, onion_session_privs) = self.create_pending_payment(
+					let (retryable_payment, onion_session_privs) = Self::create_pending_payment(
 						payment_hash, recipient_onion.clone(), keysend_preimage, None, &route,
 						Some(retry_strategy), payment_params, entropy_source, best_block_height
 					);
@@ -965,7 +965,7 @@ impl OutboundPayments {
 					let invreq = if let PendingOutboundPayment::StaticInvoiceReceived { invoice_request, .. } = entry.remove() {
 						invoice_request
 					} else { unreachable!() };
-					let (retryable_payment, onion_session_privs) = self.create_pending_payment(
+					let (retryable_payment, onion_session_privs) = Self::create_pending_payment(
 						payment_hash, recipient_onion.clone(), keysend_preimage, Some(invreq), &route,
 						Some(retry_strategy), payment_params, entropy_source, best_block_height
 					);
@@ -1572,7 +1572,7 @@ impl OutboundPayments {
 		match pending_outbounds.entry(payment_id) {
 			hash_map::Entry::Occupied(_) => Err(PaymentSendFailure::DuplicatePayment),
 			hash_map::Entry::Vacant(entry) => {
-				let (payment, onion_session_privs) = self.create_pending_payment(
+				let (payment, onion_session_privs) = Self::create_pending_payment(
 					payment_hash, recipient_onion, keysend_preimage, None, route, retry_strategy,
 					payment_params, entropy_source, best_block_height
 				);
@@ -1583,7 +1583,7 @@ impl OutboundPayments {
 	}
 
 	fn create_pending_payment<ES: Deref>(
-		&self, payment_hash: PaymentHash, recipient_onion: RecipientOnionFields,
+		payment_hash: PaymentHash, recipient_onion: RecipientOnionFields,
 		keysend_preimage: Option<PaymentPreimage>, invoice_request: Option<InvoiceRequest>,
 		route: &Route, retry_strategy: Option<Retry>, payment_params: Option<PaymentParameters>,
 		entropy_source: &ES, best_block_height: u32
@@ -2276,8 +2276,8 @@ impl_writeable_tlv_based_enum_upgradable!(PendingOutboundPayment,
 		(2, retry_strategy, required),
 		(4, max_total_routing_fee_msat, option),
 	},
-	// Added in 0.0.125. Prior versions will drop these outbounds on downgrade, which is safe because
-	// no HTLCs are in-flight.
+	// Added in 0.1. Prior versions will drop these outbounds on downgrade, which is safe because no
+	// HTLCs are in-flight.
 	(9, StaticInvoiceReceived) => {
 		(0, payment_hash, required),
 		(2, keysend_preimage, required),
```

I also removed the links to https://github.com/lightning/bolts/pull/1149 in the commit messages because I realized they were majorly cluttering up that PR.
